File: NetFramework/DataFilter/CS/ChartAsDataFilter/Extensions/DateTimeExtension.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ChartAsDataFilter
{
    public static class DateTimeExtensions
    {
        static GregorianCalendar _gc = new GregorianCalendar();
        public static int GetQuarterNumber(this DateTime date)
        {
            return (date.Date.Month - 1) / 3 + 1;
        }
        public static DateTime GetFirstDateOfQuarter(this DateTime date)
        {
            return new DateTime(date.Date.Year, (date.GetQuarterNumber() - 1) * 3 + 1, 1);
        }
        public static DateTime GetLastDateOfQuarter(this DateTime date)
        {
            return date.GetFirstDateOfQuarter().AddMonths(3).AddDays(-1);
        }
        public static int GetWeekOfMonth(this DateTime date)
        {
            var first = new DateTime(date.Year, date.Month, 1);
            return date.GetWeekOfYear() - first.GetWeekOfYear() + 1;
        }
        public static DateTime GetFirstDateOfWeek(this DateTime date)
        {
            var monthFirst = new DateTime(date.Year, date.Month, 1);
            var offset = (int)CultureInfo.CurrentCulture.DateTimeFormat.FirstDayOfWeek - (int)monthFirst.DayOfWeek;
            var weekNum = date.GetWeekOfMonth();
            return weekNum==1?monthFirst: monthFirst.AddDays((weekNum - 1) * 7 + offset);
        }
        public static DateTime GetLastDateOfWeek(this DateTime date)
        {
            var firstDate = date.GetFirstDateOfWeek();
            var lastDate = firstDate.AddDays(6 - (int)firstDate.DayOfWeek);
            return lastDate.Month != date.Month ? new DateTime(date.Year, date.Month, DateTime.DaysInMonth(date.Year, date.Month)) : lastDate;
        }
        public static int GetWeekOfYear(this DateTime date)
        {
            return _gc.GetWeekOfYear(date, CalendarWeekRule.FirstDay, DayOfWeek.Monday);
        }
    }
}

File: NetFramework/DataFilter/VB/ChartAsDataFilter/Extensions/DateTimeExtension.vb
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Runtime.CompilerServices

Namespace ChartAsDataFilter
    Module DateTimeExtensions
        Dim _gc As GregorianCalendar = New GregorianCalendar()

        <Extension()>
        Function GetQuarterNumber(ByVal day As DateTime) As Integer
            Return (day.Date.Month - 1) / 3 + 1
        End Function

        <Extension()>
        Function GetFirstDateOfQuarter(ByVal day As DateTime) As DateTime
            Return New DateTime(day.Date.Year, (day.GetQuarterNumber() - 1) * 3 + 1, 1)
        End Function

        <Extension()>
        Function GetLastDateOfQuarter(ByVal day As DateTime) As DateTime
            Return day.GetFirstDateOfQuarter().AddMonths(3).AddDays(-1)
        End Function

        <Extension()>
        Function GetWeekOfMonth(ByVal day As DateTime) As Integer
            Dim first = New DateTime(day.Year, day.Month, 1)
            Return day.GetWeekOfYear() - first.GetWeekOfYear() + 1
        End Function

        <Extension()>
        Function GetFirstDateOfWeek(ByVal day As DateTime) As DateTime
            Dim monthFirst = New DateTime(day.Year, day.Month, 1)
            Dim offset = CInt(CultureInfo.CurrentCulture.DateTimeFormat.FirstDayOfWeek) - CInt(monthFirst.DayOfWeek)
            Dim weekNum = day.GetWeekOfMonth()
            Return If(weekNum = 1, monthFirst, monthFirst.AddDays((weekNum - 1) * 7 + offset))
        End Function

        <Extension()>
        Function GetLastDateOfWeek(ByVal day As DateTime) As DateTime
            Dim firstDate = day.GetFirstDateOfWeek()
            Dim lastDate = firstDate.AddDays(6 - CInt(firstDate.DayOfWeek))
            Return If(lastDate.Month <> day.Month, New DateTime(day.Year, day.Month, DateTime.DaysInMonth(day.Year, day.Month)), lastDate)
        End Function

        <Extension()>
        Function GetWeekOfYear(ByVal day As DateTime) As Integer
            Return _gc.GetWeekOfYear(day, CalendarWeekRule.FirstDay, DayOfWeek.Monday)
        End Function
    End Module
End Namespace